### ISSUE TITLE
Issues with the train_semisupervised_seg script and question about synthmorph

### ISSUE THREAD

**Issue by eyalhana** [voxelmorph/voxelmorph#496]
First of all, thanks for the great package!!

### 1. train_semisupervised_seg
I would appreciate assistance regarding the use of semisupervised_seg script.

I'm trying to understand what the output from the generator should be.
I create my own generator and not using the original.

I have two-dimensional image data with the size of (W,D) and a batch size of BS.
I also Have segmentations which include one label (So they are binaries masks with '1' for the foreground and '0' for the background.

According to the tutorial for an unsupervised case without segmentation the output should be

```
# unsupervised 2D case
in_sample, out_sample = next(train_generator)

# shapes:
# in_sample[0].shape = (BS,W,H,1)
# in_sample[1].shape = (BS,W,H,1)
# out_sample[0].shape = (BS,W,H,1)
# out_sample[1].shape = (BS,W,H,2)
# inshape = (160,160)
```

The semisupervised case does not appear in the tutorial. 
I thought the shapes of the output of the generator should be:

```
# semi_supervised 2D case with the default downsize of 2
#in_sample[0].shape = (BS,W,H,1).         #src_vol
#in_sample[1].shape = (BS,W,H,1).          #trg_vol
#in_sample[2].shape = (BS,W/2,H/2,1).   #src_seg
#out_sample[0].shape = (BS,W,H,1).       #trg_vol
#out_sample[1].shape = (BS,W,H,2).       # zeros
#out_sample[2].shape = (BS,W/2,H/2,1)  #trg_seg
#inshape = (160,160)
```

But I get an error:
```
Exception has occurred: InvalidArgumentError       (note: full exception trace is shown but execution is paused at: _run_module_as_main)
Graph execution error:

Node 'training/Adam/gradients/gradients/vxm_dense_transformer/map/while_grad/vxm_dense_transformer/map/while_grad': Connecting to invalid output 154 of source node vxm_dense_transformer/map/while which has 154 outputs. Try using tf.compat.v1.experimental.output_all_intermediates(True)
```
Did I use the wrong dimensions? what should I do? 

### 2. synthmorh
another issue - I want to use synthmorph for my dataset. I follow the synthmorph tutorial and it works great, but I do not used segmentations. 
I wonder how should I create synthetic data that use the segmentations (like you did in the paper).
I do not find where you did it in the code/tutorial ( I saw just the non-segmentation case).

Thanks you very much!!!

**Comment by adalca** [comment]
I'll tag @mu40 for the synthmorph, and @ahoopes might chime in about the semi-supervised.
Can you tell us what line gives you that error, is it building the model or doing model.fit? Anyway when you build the model, you can print model.inputs and model.outputs to see the required sizes, I believe?

**Comment by eyalhana** [comment]
thank you very much for the quick answer!

As you advised, I checked the required sizes with the help of model.outputs and model.inputs.

I found out that I did have a mistake in the dimensions.

The size of out_sample[1] should have been (BS,W/2,H/2,2) and not (BS,W,H,2).

But still I get the same error in the model.fit_generator.

I'm not sure the message I received is informative. This is the message:

```
Epoch 1/1500
2023-01-08 22:48:20.598717: W tensorflow/c/c_api.cc:349] Operation '{name:'vxm_dense_flow_resize/map/while' id:772 
op device:{requested: '', assigned: ''} def:{{{node vxm_dense_flow_resize/map/while}} = StatelessWhile[T=[DT_INT32, 
DT_INT32, DT_INT32, DT_VARIANT, DT_INT32, 7950675786703741945, DT_VARIANT, DT_VARIANT, DT_VARIANT, 
DT_VARIANT, DT_VARIANT], _lower_using_switch_merge=true, _num_original_outputs=251, _read_only_resource_inputs=
[], _stateful_parallelism=false, body=vxm_dense_flow_resize_map_while_body_261_rewritten[], 
cond=vxm_dense_flow_resize_map_while_cond_260_rewritten[], output_shapes=[[], [], [], [], [], 16631916228386818743, 
[], [], [], [], []], parallel_iterations=10](vxm_dense_flow_resize/map/while/loop_counter, 
vxm_dense_flow_resize/map/strided_slice, vxm_dense_flow_resize/map/Const, 
vxm_dense_flow_resize/map/TensorArrayV2_1, vxm_dense_flow_resize/map/strided_slice, 
vxm_dense_flow_resize/map/TensorArrayUnstack/TensorListFromTensor, 
vxm_dense_flow_resize/map/while/EmptyTensorList, vxm_dense_flow_resize/map/while/EmptyTensorList_1, 
vxm_dense_flow_resize/map/while/EmptyTensorList_2, vxm_dense_flow_resize/map/while/EmptyTensorList_3, 
vxm_dense_flow_resize/map/while/EmptyTensorList_4, vxm_dense_flow_resize/map/while/EmptyTensorList_5, 
vxm_dense_flow_resize/map/while/EmptyTensorList_6, vxm_dense_flow_resize/map/while/EmptyTensorList_7, 
vxm_dense_flow_resize/map/while/EmptyTensorList_8, vxm_dense_flow_resize/map/while/EmptyTensorList_9, vxm_dense_flow_resize/map/while/EmptyTensorList_10, vxm_dense_flow_resize/map/while/EmptyTensorList_11, vxm_dense_flow_resize/map/while/EmptyTensorList_12, vxm_dense_flow_resize/map/while/EmptyTensorList_13, 
vxm_dense_flow_resize/map/while/EmptyTensorList_14, vxm_dense_flow_resize/map/while/EmptyTensorList_15, 
vxm_dense_flow_resize/map/while/EmptyTensorList_16, vxm_dense_flow_resize/map/while/EmptyTensorList_17, 
vxm_dense_flow_resize/map/while/EmptyTensorList_18, vxm_dense_flow_resize/map/while/EmptyTensorList_19, 
vxm_dense_flow_resize/map/while/EmptyTensorList_20, vxm_dense_flow_resize/map/while/EmptyTensorList_21, 
vxm_dense_flow_resize/map/while/EmptyTensorList_22, vxm_dense_flow_resize/map/while/EmptyTensorList_23, 
vxm_dense_flow_resize/map/while/EmptyTensorList_24, vxm_dense_flow_resize/map/while/EmptyTensorList_25, 
vxm_dense_flow_resize/map/while/EmptyTensorList_26, vxm_dense_flow_resize/map/while/EmptyTensorList_27, 
vxm_dense_flow_resize/map/while/EmptyTensorList_28, vxm_dense_flow_resize/map/while/EmptyTensorList_29, 
vxm_dense_flow_resize/map/while/EmptyTensorList_30, vxm_dense_flow_resize/map/while/EmptyTensorList_31, 
vxm_dense_flow_resize/map/while/EmptyTensorList_32, vxm_dense_flow_resize/map/while/EmptyTensorList_33, 
vxm_dense_flow_resize/map/while/EmptyTensorList_34, vxm_dense_flow_resize/map/while/EmptyTensorList_35, 
vxm_dense_flow_resize/map/while/EmptyTensorList_36, vxm_dense_flow_resize/map/while/EmptyTensorList_37, 
vxm_dense_flow_resize/map/while/EmptyTensorList_38, vxm_dense_flow_resize/map/while/EmptyTensorList_39, 
vxm_dense_flow_resize/map/while/EmptyTensorList_40, vxm_dense_flow_resize/map/while/EmptyTensorList_41, 
vxm_dense_flow_resize/map/while/EmptyTensorList_42, vxm_dense_flow_resize/map/while/EmptyTensorList_43, 
vxm_dense_flow_resize/map/while/EmptyTensorList_44, vxm_dense_flow_resize/map/while/EmptyTensorList_45, 
vxm_dense_flow_resize/map/while/EmptyTensorList_46, vxm_dense_flow_resize/map/while/EmptyTensorList_47, 
vxm_dense_flow_resize/map/while/EmptyTensorList_48, vxm_dense_flow_resize/map/while/EmptyTensorList_49, 
vxm_dense_flow_resize/map/while/EmptyTensorList_50, vxm_dense_flow_resize/map/while/EmptyTensorList_51, 
vxm_dense_flow_resize/map/while/EmptyTensorList_52, vxm_dense_flow_resize/map/while/EmptyTensorList_53, 
vxm_dense_flow_resize/map/while/EmptyTensorList_54, vxm_dense_flow_resize/map/while/EmptyTensorList_55, 
vxm_dense_flow_resize/map/while/EmptyTensorList_56, vxm_dense_flow_resize/map/while/EmptyTensorList_57, 
vxm_dense_flow_resize/map/while/EmptyTensorList_58, vxm_dense_flow_resize/map/while/EmptyTensorList_59, 
vxm_dense_flow_resize/map/while/EmptyTensorList_60, vxm_dense_flow_resize/map/while/EmptyTensorList_61, 
vxm_dense_flow_resize/map/while/EmptyTensorList_62, vxm_dense_flow_resize/map/while/EmptyTensorList_63, 
vxm_dense_flow_resize/map/while/EmptyTensorList_64, vxm_dense_flow_resize/map/while/EmptyTensorList_65, 
vxm_dense_flow_resize/map/while/EmptyTensorList_66, vxm_dense_flow_resize/map/while/EmptyTensorList_67, 
vxm_dense_flow_resize/map/while/EmptyTensorList_68, vxm_dense_flow_resize/map/while/EmptyTensorList_69, 
vxm_dense_flow_resize/map/while/EmptyTensorList_70, vxm_dense_flow_resize/map/while/EmptyTensorList_71, 
vxm_dense_flow_resize/map/while/EmptyTensorList_72, vxm_dense_flow_resize/map/while/EmptyTensorList_73, 
vxm_dense_flow_resize/map/while/EmptyTensorList_74, vxm_dense_flow_resize/map/while/EmptyTensorList_75, 
vxm_dense_flow_resize/map/while/EmptyTensorList_76, vxm_dense_flow_resize/map/while/EmptyTensorList_77, 
vxm_dense_flow_resize/map/while/EmptyTensorList_78, vxm_dense_flow_resize/map/while/EmptyTensorList_79, 
vxm_dense_flow_resize/map/while/EmptyTensorList_80, vxm_dense_flow_resize/map/while/EmptyTensorList_81, 
vxm_dense_flow_resize/map/while/EmptyTensorList_82, vxm_dense_flow_resize/map/while/EmptyTensorList_83, 
vxm_dense_flow_resize/map/while/EmptyTensorList_84, vxm_dense_flow_resize/map/while/EmptyTensorList_85, 
vxm_dense_flow_resize/map/while/EmptyTensorList_86, vxm_dense_flow_resize/map/while/EmptyTensorList_87, 
vxm_dense_flow_resize/map/while/EmptyTensorList_88, vxm_dense_flow_resize/map/while/EmptyTensorList_89, 
vxm_dense_flow_resize/map/while/EmptyTensorList_90, vxm_dense_flow_resize/map/while/EmptyTensorList_91, 
vxm_dense_flow_resize/map/while/EmptyTensorList_92, vxm_dense_flow_resize/map/while/EmptyTensorList_93, 
vxm_dense_flow_resize/map/while/EmptyTensorList_94, vxm_dense_flow_resize/map/while/EmptyTensorList_95, 
vxm_dense_flow_resize/map/while/EmptyTensorList_96, vxm_dense_flow_resize/map/while/EmptyTensorList_97, 
vxm_dense_flow_resize/map/while/EmptyTensorList_98, vxm_dense_flow_resize/map/while/EmptyTensorList_99, 
vxm_dense_flow_resize/map/while/EmptyTensorList_100, vxm_dense_flow_resize/map/while/EmptyTensorList_101, 
vxm_dense_flow_resize/map/while/EmptyTensorList_102, vxm_dense_flow_resize/map/while/EmptyTensorList_103, 
vxm_dense_flow_resize/map/while/EmptyTensorList_104, vxm_dense_flow_resize/map/while/EmptyTensorList_105, 
vxm_dense_flow_resize/map/while/EmptyTensorList_106, vxm_dense_flow_resize/map/while/EmptyTensorList_107, 
vxm_dense_flow_resize/map/while/EmptyTensorList_108, vxm_dense_flow_resize/map/while/EmptyTensorList_109, 
vxm_dense_flow_resize/map/while/EmptyTensorList_110, vxm_dense_flow_resize/map/while/EmptyTensorList_111, 
vxm_dense_flow_resize/map/while/EmptyTensorList_112, vxm_dense_flow_resize/map/while/EmptyTensorList_113, 
vxm_dense_flow_resize/map/while/EmptyTensorList_114, vxm_dense_flow_resize/map/while/EmptyTensorList_115, 
vxm_dense_flow_resize/map/while/EmptyTensorList_116, vxm_dense_flow_resize/map/while/EmptyTensorList_117, 
vxm_dense_flow_resize/map/while/EmptyTensorList_118, vxm_dense_flow_resize/map/while/EmptyTensorList_119, 
vxm_dense_flow_resize/map/while/EmptyTensorList_120, vxm_dense_flow_resize/map/while/EmptyTensorList_121, 
vxm_dense_flow_resize/map/while/EmptyTensorList_122, vxm_dense_flow_resize/map/while/EmptyTensorList_123, 
vxm_dense_flow_resize/map/while/EmptyTensorList_124, vxm_dense_flow_resize/map/while/EmptyTensorList_125, 
vxm_dense_flow_resize/map/while/EmptyTensorList_126, vxm_dense_flow_resize/map/while/EmptyTensorList_127, 
vxm_dense_flow_resize/map/while/EmptyTensorList_128, vxm_dense_flow_resize/map/while/EmptyTensorList_129, 
vxm_dense_flow_resize/map/while/EmptyTensorList_130, vxm_dense_flow_resize/map/while/EmptyTensorList_131, 
vxm_dense_flow_resize/map/while/EmptyTensorList_132, vxm_dense_flow_resize/map/while/EmptyTensorList_133, 
vxm_dense_flow_resize/map/while/EmptyTensorList_134, vxm_dense_flow_resize/map/while/EmptyTensorList_135, 
vxm_dense_flow_resize/map/while/EmptyTensorList_136, vxm_dense_flow_resize/map/while/EmptyTensorList_137, 
vxm_dense_flow_resize/map/while/EmptyTensorList_138, vxm_dense_flow_resize/map/while/EmptyTensorList_139, 
vxm_dense_flow_resize/map/while/EmptyTensorList_140, vxm_dense_flow_resize/map/while/EmptyTensorList_141, 
vxm_dense_flow_resize/map/while/EmptyTensorList_142, vxm_dense_flow_resize/map/while/EmptyTensorList_143, 
vxm_dense_flow_resize/map/while/EmptyTensorList_144, vxm_dense_flow_resize/map/while/EmptyTensorList_145, 
vxm_dense_flow_resize/map/while/EmptyTensorList_146, vxm_dense_flow_resize/map/while/EmptyTensorList_147, 
vxm_dense_flow_resize/map/while/EmptyTensorList_148, vxm_dense_flow_resize/map/while/EmptyTensorList_149, 
vxm_dense_flow_resize/map/while/EmptyTensorList_150, vxm_dense_flow_resize/map/while/EmptyTensorList_151, 
vxm_dense_flow_resize/map/while/EmptyTensorList_152, vxm_dense_flow_resize/map/while/EmptyTensorList_153, 
vxm_dense_flow_resize/map/while/EmptyTensorList_154, vxm_dense_flow_resize/map/while/EmptyTensorList_155, 
vxm_dense_flow_resize/map/while/EmptyTensorList_156, vxm_dense_flow_resize/map/while/EmptyTensorList_157, 
vxm_dense_flow_resize/map/while/EmptyTensorList_158, vxm_dense_flow_resize/map/while/EmptyTensorList_159, 
vxm_dense_flow_resize/map/while/EmptyTensorList_160, vxm_dense_flow_resize/map/while/EmptyTensorList_161, 
vxm_dense_flow_resize/map/while/EmptyTensorList_162, vxm_dense_flow_resize/map/while/EmptyTensorList_163, 
vxm_dense_flow_resize/map/while/EmptyTensorList_164, vxm_dense_flow_resize/map/while/EmptyTensorList_165, 
vxm_dense_flow_resize/map/while/EmptyTensorList_166, vxm_dense_flow_resize/map/while/EmptyTensorList_167, 
vxm_dense_flow_resize/map/while/EmptyTensorList_168, vxm_dense_flow_resize/map/while/EmptyTensorList_169, 
vxm_dense_flow_resize/map/while/EmptyTensorList_170, vxm_dense_flow_resize/map/while/EmptyTensorList_171, 
vxm_dense_flow_resize/map/while/EmptyTensorList_172, vxm_dense_flow_resize/map/while/EmptyTensorList_173, 
vxm_dense_flow_resize/map/while/EmptyTensorList_174, vxm_dense_flow_resize/map/while/EmptyTensorList_175, 
vxm_dense_flow_resize/map/while/EmptyTensorList_176, vxm_dense_flow_resize/map/while/EmptyTensorList_177, 
vxm_dense_flow_resize/map/while/EmptyTensorList_178, vxm_dense_flow_resize/map/while/EmptyTensorList_179, 
vxm_dense_flow_resize/map/while/EmptyTensorList_180, vxm_dense_flow_resize/map/while/EmptyTensorList_181, 
vxm_dense_flow_resize/map/while/EmptyTensorList_182, vxm_dense_flow_resize/map/while/EmptyTensorList_183, 
vxm_dense_flow_resize/map/while/EmptyTensorList_184, vxm_dense_flow_resize/map/while/EmptyTensorList_185, 
vxm_dense_flow_resize/map/while/EmptyTensorList_186, vxm_dense_flow_resize/map/while/EmptyTensorList_187, 
vxm_dense_flow_resize/map/while/EmptyTensorList_188, vxm_dense_flow_resize/map/while/EmptyTensorList_189, 
vxm_dense_flow_resize/map/while/EmptyTensorList_190, vxm_dense_flow_resize/map/while/EmptyTensorList_191, 
vxm_dense_flow_resize/map/while/EmptyTensorList_192, vxm_dense_flow_resize/map/while/EmptyTensorList_193, 
vxm_dense_flow_resize/map/while/EmptyTensorList_194, vxm_dense_flow_resize/map/while/EmptyTensorList_195, 
vxm_dense_flow_resize/map/while/EmptyTensorList_196, vxm_dense_flow_resize/map/while/EmptyTensorList_197, 
vxm_dense_flow_resize/map/while/EmptyTensorList_198, vxm_dense_flow_resize/map/while/EmptyTensorList_199, 
vxm_dense_flow_resize/map/while/EmptyTensorList_200, vxm_dense_flow_resize/map/while/EmptyTensorList_201, vxm_dense_flow_resize/map/while/EmptyTensorList_202, vxm_dense_flow_resize/map/while/EmptyTensorList_203, 
vxm_dense_flow_resize/map/while/EmptyTensorList_204, vxm_dense_flow_resize/map/while/EmptyTensorList_205, 
vxm_dense_flow_resize/map/while/EmptyTensorList_206, vxm_dense_flow_resize/map/while/EmptyTensorList_207, 
vxm_dense_flow_resize/map/while/EmptyTensorList_208, vxm_dense_flow_resize/map/while/EmptyTensorList_209, 
vxm_dense_flow_resize/map/while/EmptyTensorList_210, vxm_dense_flow_resize/map/while/EmptyTensorList_211, 
vxm_dense_flow_resize/map/while/EmptyTensorList_212, vxm_dense_flow_resize/map/while/EmptyTensorList_213, 
vxm_dense_flow_resize/map/while/EmptyTensorList_214, vxm_dense_flow_resize/map/while/EmptyTensorList_215, 
vxm_dense_flow_resize/map/while/EmptyTensorList_216, vxm_dense_flow_resize/map/while/EmptyTensorList_217, 
vxm_dense_flow_resize/map/while/EmptyTensorList_218, vxm_dense_flow_resize/map/while/EmptyTensorList_219, 
vxm_dense_flow_resize/map/while/EmptyTensorList_220, vxm_dense_flow_resize/map/while/EmptyTensorList_221, 
vxm_dense_flow_resize/map/while/EmptyTensorList_222, vxm_dense_flow_resize/map/while/EmptyTensorList_223, 
vxm_dense_flow_resize/map/while/EmptyTensorList_224, vxm_dense_flow_resize/map/while/EmptyTensorList_225, 
vxm_dense_flow_resize/map/while/EmptyTensorList_226, vxm_dense_flow_resize/map/while/EmptyTensorList_227, 
vxm_dense_flow_resize/map/while/EmptyTensorList_228, vxm_dense_flow_resize/map/while/EmptyTensorList_229, 
vxm_dense_flow_resize/map/while/EmptyTensorList_230, vxm_dense_flow_resize/map/while/EmptyTensorList_231, 
vxm_dense_flow_resize/map/while/EmptyTensorList_232, vxm_dense_flow_resize/map/while/EmptyTensorList_233, 
vxm_dense_flow_resize/map/while/EmptyTensorList_234, vxm_dense_flow_resize/map/while/EmptyTensorList_235, 
vxm_dense_flow_resize/map/while/EmptyTensorList_236, vxm_dense_flow_resize/map/while/EmptyTensorList_237, 
vxm_dense_flow_resize/map/while/EmptyTensorList_238, vxm_dense_flow_resize/map/while/EmptyTensorList_239, 
vxm_dense_flow_resize/map/while/EmptyTensorList_240, vxm_dense_flow_resize/map/while/EmptyTensorList_241, 
vxm_dense_flow_resize/map/while/EmptyTensorList_242, 
training/Adam/gradients/gradients/vxm_dense_flow_resize/map/while_grad/gradients/TensorArrayV2Read/TensorListGetItem_grad/TensorListElementShape_0/accumulator:0, 
training/Adam/gradients/gradients/vxm_dense_flow_resize/map/while_grad/gradients/TensorArrayV2Read/TensorListGetItem_grad/TensorListLength_0/accumulator:0)}}' 
was changed by setting attribute after it was run by a session. 
This mutation will have no effect, and will trigger an error in the future. 
Either don't modify nodes after running them or create a new session.
```

**Comment by adalca** [comment]
Hm I haven't seen this, It might help if you tell us a bit more of your setup/tf-versions/etc.
And Malte/Andrew might also have seen more of these errors

**Comment by mu40** [comment]
Yes, it would be helpful to know how you set up your environment. I would generally recommend using a recent TensorFlow version for our VoxelMorph/SynthMorph.

Regarding your question about SynthMorph, are you asking how to train a network using your own anatomical label maps or how to create those label maps from brain MRI scans, for example? If it is the former, you can feed `vxm.generators.synthmorph` in the tutorial your own label maps, provided that you created the registration model with the appropriate shape.

**Comment by eyalhana** [comment]
Thanks a lot for the advice!!! I appreciate your quick response!

I reinstalled TensorFlow and changed the version to the same version in the tutorial (2.9.2).

For some reason, the problem continued, so I added the following line:
```tf.compat.v1.experimental.output_all_intermediates(True)```

I don't know what it does, but the problem is solved (I came across it from a random StackOverflow search).

I have a follow-up question - I am interested in using mutual information as a loss.
In the semisupervised_seg code itself, there is no MI flag, so it has to be added manually.
I wonder if, by any chance, you remember if I should also change the hyperparameters (for example, the LR, lambda, gamma) or to keep the default hyperparameters of the mse/ncc cases .

As for the synthmorph, I followed @mu40 answer, and it worked!

Thanks again @adalca  @mu40 for all the help :-)
